README.md
## NCM Dump
Decrypt .ncm file and convert it to normal music format.
### Usage
```dart
final ncm = NCM();

final raw = await File('a.ncm').readAsBytes();
ncm.parse(raw);
await File('a.${ncm.meta.format}').writeAsBytes(ncm.music);
```


### Test
Please copy your .ncm file to `test/assets/a.ncm`.  
And run the following command to test.
```bash
# convert .ncm to common music format
python3 test/ncmdumpy.py test/assets/a.ncm -o
# run dart test
dart test
```


CHANGELOG.md
## 0.0.1
* Support parse .ncm file.

## 0.0.2
* Fix NCM._readMusic() chunk length error.

## 0.0.3
* Add comment

## 0.0.4
- Remove setRaw() function. Use parse() function instead.


test/ncmdump.py
# -*- coding: utf-8 -*-
# https://github.com/LollipopKit/ncmdump

import binascii
import struct
import base64
import json
import os
import sys
from Crypto.Cipher import AES

NCM_SUFFIX = '.ncm'
AUDIO_SUFFIX = ['.mp3', '.flac']


def unpad(s):
    return s[0:-(s[-1] if type(s[-1]) == int else ord(s[-1]))]


def dump(file_path):
    # AES key
    core_key = binascii.a2b_hex("687A4852416D736F356B496E62617857")
    meta_key = binascii.a2b_hex("2331346C6A6B5F215C5D2630553C2728")

    # read magic header
    f = open(file_path, 'rb')
    header = f.read(8)
    # str to hex
    assert binascii.b2a_hex(header) == b'4354454e4644414d'

    # read key length
    f.seek(2, 1)
    key_length = f.read(4)
    key_length = struct.unpack('<I', bytes(key_length))[0]

    # read key
    key_data = f.read(key_length)
    key_data_array = bytearray(key_data)
    for i in range(0, len(key_data_array)):
        key_data_array[i] ^= 0x64
    key_data = bytes(key_data_array)
    cryptor = AES.new(core_key, AES.MODE_ECB)
    key_data = unpad(cryptor.decrypt(key_data))[17:]
    key_length = len(key_data)
    key_data = bytearray(key_data)
    key_box = bytearray(range(256))
    c = 0
    last_byte = 0
    key_offset = 0
    for i in range(256):
        swap = key_box[i]
        c = (swap + last_byte + key_data[key_offset]) & 0xff
        key_offset += 1
        if key_offset >= key_length:
            key_offset = 0
        key_box[i] = key_box[c]
        key_box[c] = swap
        last_byte = c

    # read meta length
    meta_length = f.read(4)
    meta_length = struct.unpack('<I', bytes(meta_length))[0]

    # read meta
    meta_data = f.read(meta_length)
    meta_data_array = bytearray(meta_data)
    for i in range(0, len(meta_data_array)):
        meta_data_array[i] ^= 0x63
    meta_data = bytes(meta_data_array)
    meta_data = base64.b64decode(meta_data[22:])
    cryptor = AES.new(meta_key, AES.MODE_ECB)
    meta_data = unpad(cryptor.decrypt(meta_data)).decode('utf-8')[6:]
    meta_data = json.loads(meta_data)

    # read crc
    crc32 = f.read(4)
    crc32 = struct.unpack('<I', bytes(crc32))[0]
    f.seek(5, 1)

    # read image size
    image_size = f.read(4)
    image_size = struct.unpack('<I', bytes(image_size))[0]

    # read image data
    f.read(image_size)

    # read music data
    file_name = f.name.split("/")[-1].replace('.ncm', f".{meta_data['format']}")
    m = open(os.path.join(os.path.split(file_path)[0], file_name), 'wb')
    chunk = bytearray()
    while True:
        chunk = bytearray(f.read(0x8000))
        if not chunk:
            break
        for i in range(1, len(chunk)+1):
            j = i & 0xff
            chunk[i-1] ^= key_box[(key_box[j] + key_box[(key_box[j] + j) & 0xff]) & 0xff]
        m.write(chunk)
    m.close()
    f.close()
    return file_name


if __name__ == '__main__':
    args = sys.argv[1:]

    if not args:
        print('Usage: \n  python3 ncmdump.py <ncm_folder_path>\n  Add "-o" to override.')
        exit(1)

    folder = args[0]
    override = '-o' == args[1] if len(args) > 1 else False

    file_list = []
    for file in os.listdir(folder):
        if file.endswith(NCM_SUFFIX):
            if not override:
                have = False
                for suffix in AUDIO_SUFFIX:
                    if os.path.exists(os.path.join(folder, file.replace(NCM_SUFFIX, suffix))):
                        have = True
                        break
                if have:
                    continue
            file_list.append(os.path.join(folder, file))

    convert_length = len(file_list)
    if convert_length == 0:
        print(f"No .ncm files need to be converted.\nUse 'python3 ncmdump.py {folder} -o' to override.")
        exit(0)

    for index, file in enumerate(file_list):
        try:
            print(f'[{index}/{convert_length}]  {dump(file)}')
        except KeyboardInterrupt as e:
            if 'm.write(chunk)' in str(e):
                for suffix in AUDIO_SUFFIX:
                    file_path = os.path.join(folder, file.replace(NCM_SUFFIX, suffix))
                    if os.path.exists(file_path):
                        os.remove(file_path)
                        print(f'\nDue to KeyboardInterrupt\n{file_path} is removed.')
            exit(0)


test/ncmdump_test.dart
import 'dart:io';

import 'package:ncmdump/ncmdump.dart';
import 'package:test/test.dart';

const ncmPath = 'test/assets/a.ncm';
const mp3Path = 'test/assets/a.mp3';

void main() {
  test('convert .ncm', () async {
    final ncmFile = File(ncmPath);
    final mp3File = File(mp3Path);
    if (!ncmFile.existsSync() || !mp3File.existsSync()) {
      print(
          'please put your own test .ncm file in "$ncmPath", and run "python3 ncmdump.py \'test/assets\'"');
      return;
    }
    final raw = await ncmFile.readAsBytes();
    final target = await mp3File.readAsBytes();
    final ncm = NCM();
    ncm.parse(raw);
    // ignore: avoid_print
    print(
        'name: ${ncm.meta.musicName}, format: ${ncm.meta.format}, artist: ${ncm.meta.artist}');
    expect(ncm.music, target);
  });
}


lib/src/model/ncm.dart
// ignore_for_file: avoid_print

import 'dart:convert';
import 'dart:typed_data';

import 'package:encrypt/encrypt.dart';
import 'package:ncmdump/src/constant.dart';
import 'package:ncmdump/src/ext/convert.dart';
import 'package:ncmdump/src/model/meta.dart';

/// Construct a new NCM instance.
/// ## Example:
/// ```dart
/// final ncm = NCM();
/// final raw = await File('a.ncm').readAsBytes();
/// ncm.parse(raw);
/// await File('a.${ncm.meta.format}').writeAsBytes(ncm.music);
/// ```
class NCM {
  Uint8List? _raw;
  late String _magicHeader;
  late int _keyLength;
  late Uint8List key;
  late int _metaLength;
  late Meta meta;
  late int _imageSize;
  late Uint8List image;
  late Uint8List music;
  final Encrypter _coreCryptor;
  final Encrypter _metaCryptor;
  final IV _iv;

  NCM()
      : _iv = IV.fromLength(0),
        _coreCryptor =
            Encrypter(AES(Key.fromUtf8(coreKeyStr), mode: AESMode.ecb)),
        _metaCryptor = Encrypter(
            AES(Key.fromUtf8(hexParse(metaKeyHex)), mode: AESMode.ecb));

  /// Parse raw ncm data.
  void parse(Uint8List raw) {
    _raw = raw;
    try {
      _magicHeader = _readMagicHeader();
      assert(_magicHeader == magicHeaderValue);
      _keyLength = _readKeyLength();
      key = _readKey();
      _metaLength = _readMetaLength();
      meta = _readMeta();
      _readCRC();
      _readGap();
      _imageSize = _readImageSize();
      image = _readImage();
      music = _readMusic();
    } catch (e) {
      print('parse failed: $e');
      rethrow;
    }
  }

  String _readMagicHeader() {
    final data = _raw!.sublist(0, 8);
    _raw = _raw!.sublist(10);
    return data.hex;
  }

  int _readKeyLength() {
    final data = _raw!.sublist(0, 4);
    _raw = _raw!.sublist(4);
    return data.getLength();
  }

  Uint8List _readKey() {
    final data = _raw!.sublist(0, _keyLength);
    _raw = _raw!.sublist(_keyLength);

    /// CR4-KSA
    for (var i = 0; i < data.length; i++) {
      data[i] ^= 0x64;
    }
    final aesDecrypted = _coreCryptor.decrypt(Encrypted(data), iv: _iv);
    final keyData = Uint8List.fromList(aesDecrypted.substring(17).codeUnits);
    return _decryptKey(keyData);
  }

  /// CR4-PRGA
  Uint8List _decryptKey(Uint8List raw) {
    final keyBox = List<int>.generate(256, (index) => index);
    var c = 0;
    var lastByte = 0;
    var keyOffset = 0;
    for (var i = 0; i < 256; i++) {
      var swap = keyBox[i];
      c = (swap + lastByte + raw[keyOffset]) & 0xff;
      keyOffset += 1;
      if (keyOffset >= raw.length) {
        keyOffset = 0;
      }

      keyBox[i] = keyBox[c];
      keyBox[c] = swap;
      lastByte = c;
    }

    return Uint8List.fromList(keyBox);
  }

  int _readMetaLength() {
    final data = _raw!.sublist(0, 4);
    _raw = _raw!.sublist(4);
    return data.getLength();
  }

  Meta _readMeta() {
    final data = _raw!.sublist(0, _metaLength);
    _raw = _raw!.sublist(_metaLength);

    for (var i = 0; i < _metaLength; i++) {
      data[i] ^= 0x63;
    }

    final b64Decoded = base64.decode(data.sublist(22).string);
    final aesDecrypted = _metaCryptor.decrypt(Encrypted(b64Decoded), iv: _iv);

    return Meta.fromJson(json.decode(aesDecrypted.substring(6)));
  }

  void _readCRC() {
    _raw = _raw!.sublist(4);
  }

  void _readGap() {
    _raw = _raw!.sublist(5);
  }

  int _readImageSize() {
    final data = _raw!.sublist(0, 4);
    _raw = _raw!.sublist(4);
    return data.getLength();
  }

  Uint8List _readImage() {
    final data = _raw!.sublist(0, _imageSize);
    _raw = _raw!.sublist(_imageSize);
    return data;
  }

  Uint8List _readMusic() {
    final music = List<int>.empty(growable: true);
    while (_raw!.isNotEmpty) {
      final readLength = _raw!.length < musicChunkReadLength
          ? _raw!.length
          : musicChunkReadLength;
      var chunk = _raw!.sublist(0, readLength);
      _raw = _raw!.sublist(readLength);
      for (var i = 1; i < readLength + 1; i++) {
        final j = i & 0xff;
        chunk[i - 1] ^= key[(key[j] + key[(key[j] + j) & 0xff]) & 0xff];
      }
      music.addAll(chunk);
    }
    return Uint8List.fromList(music);
  }
}


lib/src/model/meta.dart
import 'package:ncmdump/src/model/artist.dart';

///
/// Code generated by jsonToDartModel https://ashamp.github.io/jsonToDartModel/
///
class Meta {
/*
{
  "format": "flac",
  "musicId": 26112101,
  "musicName": "いつかどこかで (Album Version)",
  "artist": [
    [
      "□□□(クチロロ)", 1231231
    ]
  ],
  "album": "マンパワー",
  "albumId": 2389206,
  "albumPicDocId": 18545462627518744,
  "albumPic": "http://p1.music.126.net/xQ8-zNYuS4-6RDIubGDGSQ==/18545462627518744.jpg",
  "mvId": 0,
  "flag": 261,
  "bitrate": 1013110,
  "duration": 831000,
  "gain": -12.327300071716309,
  "alias": [
    ""
  ],
  "transNames": [
    "哪天在某处再会"
  ]
} 
*/

  String? format;
  int? musicId;
  String? musicName;
  List<Artist?>? artist;
  String? album;
  int? albumId;
  int? albumPicDocId;
  String? albumPic;
  int? mvId;
  int? flag;
  int? bitrate;
  int? duration;
  double? gain;
  List<String?>? alias;
  List<String?>? transNames;

  Meta({
    this.format,
    this.musicId,
    this.musicName,
    this.artist,
    this.album,
    this.albumId,
    this.albumPicDocId,
    this.albumPic,
    this.mvId,
    this.flag,
    this.bitrate,
    this.duration,
    this.gain,
    this.alias,
    this.transNames,
  });
  Meta.fromJson(Map<String, dynamic> json) {
    format = json['format']?.toString();
    musicId = json['musicId']?.toInt();
    musicName = json['musicName']?.toString();
    if (json['artist'] != null) {
      final v = json['artist'];
      final arr0 = <Artist>[];
      v.forEach((v) {
        arr0.add(Artist.fromJson(v));
      });
      artist = arr0;
    }
    album = json['album']?.toString();
    albumId = json['albumId']?.toInt();
    albumPicDocId = json['albumPicDocId']?.toInt();
    albumPic = json['albumPic']?.toString();
    mvId = json['mvId']?.toInt();
    flag = json['flag']?.toInt();
    bitrate = json['bitrate']?.toInt();
    duration = json['duration']?.toInt();
    gain = json['gain']?.toDouble();
    if (json['alias'] != null) {
      final v = json['alias'];
      final arr0 = <String>[];
      v.forEach((v) {
        arr0.add(v.toString());
      });
      alias = arr0;
    }
    if (json['transNames'] != null) {
      final v = json['transNames'];
      final arr0 = <String>[];
      v.forEach((v) {
        arr0.add(v.toString());
      });
      transNames = arr0;
    }
  }
  Map<String, dynamic> toJson() {
    final data = <String, dynamic>{};
    data['format'] = format;
    data['musicId'] = musicId;
    data['musicName'] = musicName;
    if (artist != null) {
      final v = artist;
      final arr0 = [];
      for (var v in v!) {
        arr0.add(v);
      }
      data['artist'] = arr0;
    }
    data['album'] = album;
    data['albumId'] = albumId;
    data['albumPicDocId'] = albumPicDocId;
    data['albumPic'] = albumPic;
    data['mvId'] = mvId;
    data['flag'] = flag;
    data['bitrate'] = bitrate;
    data['duration'] = duration;
    data['gain'] = gain;
    if (alias != null) {
      final v = alias;
      final arr0 = [];
      for (var v in v!) {
        arr0.add(v);
      }
      data['alias'] = arr0;
    }
    if (transNames != null) {
      final v = transNames;
      final arr0 = [];
      for (var v in v!) {
        arr0.add(v);
      }
      data['transNames'] = arr0;
    }
    return data;
  }
}


lib/src/ext/convert.dart
import 'dart:typed_data';

extension Uint8ListX on Uint8List {
  /// convert to hex string
  String get hex {
    var s = '';
    for (final element in this) {
      s += element.toRadixString(16);
    }
    return s;
  }

  /// convert to string
  String get string => String.fromCharCodes(this);

  /// convert 4 bytes to int (length)
  int getLength() {
    int len = 0;
    len |= this[0] & 0xff;
    len |= (this[1] & 0xff) << 8;
    len |= (this[2] & 0xff) << 16;
    len |= (this[3] & 0xff) << 24;
    return len;
  }
}

extension ByteDataX on ByteData {
  /// bytedata to string
  String get string {
    return buffer.asUint8List().string;
  }
}

/// hex string to string
String hexParse(String hex) {
  final len = hex.length ~/ 2;
  var s = '';
  for (var i = 0; i < len * 2; i += 2) {
    s += String.fromCharCode(int.tryParse(hex.substring(i, i + 2), radix: 16)!);
  }
  return s;
}


ncmdump.py
# -*- coding: utf-8 -*-
import binascii
import struct
import base64
import json
import os
import sys
from Crypto.Cipher import AES

NCM_SUFFIX = '.ncm'
AUDIO_SUFFIX = ['.mp3', '.flac']


def unpad(s):
    return s[0:-(s[-1] if type(s[-1]) == int else ord(s[-1]))]


def dump(file_path):
    # AES key
    core_key = binascii.a2b_hex("687A4852416D736F356B496E62617857")
    meta_key = binascii.a2b_hex("2331346C6A6B5F215C5D2630553C2728")

    # read magic header
    f = open(file_path, 'rb')
    header = f.read(8)
    # str to hex
    assert binascii.b2a_hex(header) == b'4354454e4644414d'

    # read key length
    f.seek(2, 1)
    key_length = f.read(4)
    key_length = struct.unpack('<I', bytes(key_length))[0]

    # read key
    key_data = f.read(key_length)
    key_data_array = bytearray(key_data)
    for i in range(0, len(key_data_array)):
        key_data_array[i] ^= 0x64
    key_data = bytes(key_data_array)

    cryptor = AES.new(core_key, AES.MODE_ECB)
    decrpted = cryptor.decrypt(key_data)

    key_data = unpad(decrpted)[17:]
    key_length = len(key_data)
    key_data = bytearray(key_data)

    key_box = bytearray(range(256))

    c = 0
    last_byte = 0
    key_offset = 0
    for i in range(256):
        swap = key_box[i]
        c = (swap + last_byte + key_data[key_offset]) & 0xff
        key_offset += 1
        if key_offset >= key_length:
            key_offset = 0
        key_box[i] = key_box[c]
        key_box[c] = swap
        last_byte = c

    key_data_int_list = []
    for i in range(0, len(key_box)):
        ii = key_box[i]
        key_data_int_list.append(ii)
    print('key_box:', key_data_int_list)
    # done

    # read meta length
    meta_length = f.read(4)
    meta_length = struct.unpack('<I', bytes(meta_length))[0]
    print('meta_length:', meta_length)

    # read meta
    meta_data = f.read(meta_length)
    meta_data_array = bytearray(meta_data)
    for i in range(0, len(meta_data_array)):
        meta_data_array[i] ^= 0x63
    meta_data = bytes(meta_data_array)
    meta_data = base64.b64decode(meta_data[22:])
    cryptor = AES.new(meta_key, AES.MODE_ECB)
    meta_data = unpad(cryptor.decrypt(meta_data)).decode('utf-8')[6:]
    meta_data = json.loads(meta_data)

    # read crc
    f.read(4)
    f.seek(5, 1)

    # read image size
    image_size = f.read(4)
    image_size = struct.unpack('<I', bytes(image_size))[0]
    print('image_size: ', image_size)

    # read image data
    f.read(image_size)

    # # read music data
    # file_name = f.name.split("/")[-1].replace('.ncm', f".{meta_data['format']}")
    # m = open(os.path.join(os.path.split(file_path)[0], file_name), 'wb')
    # chunk = bytearray()
    # while True:
    #     chunk = bytearray(f.read(0x8000))
    #     if not chunk:
    #         break
    #     for i in range(1, len(chunk)+1):
    #         j = i & 0xff
    #         chunk[i-1] ^= key_box[(key_box[j] + key_box[(key_box[j] + j) & 0xff]) & 0xff]
    #     m.write(chunk)
    # m.close()
    # f.close()
    # return file_name


if __name__ == '__main__':
    args = sys.argv[1:]

    if not args:
        print('Usage: \n  python3 ncmdump.py <ncm_folder_path>\n  Add "-o" to override.')
        exit(1)

    folder = args[0]
    override = '-o' == args[1] if len(args) > 1 else False

    file_list = []
    for file in os.listdir(folder):
        if file.endswith(NCM_SUFFIX):
            if not override:
                have = False
                for suffix in AUDIO_SUFFIX:
                    if os.path.exists(os.path.join(folder, file.replace(NCM_SUFFIX, suffix))):
                        have = True
                        break
                if have:
                    continue
            file_list.append(os.path.join(folder, file))

    convert_length = len(file_list)
    if convert_length == 0:
        print(f"No .ncm files need to be converted.\nUse 'python3 ncmdump.py {folder} -o' to override.")
        exit(0)

    for index, file in enumerate(file_list):
        try:
            print(f'[{index}/{convert_length}]  {dump(file)}')
        except KeyboardInterrupt as e:
            if 'm.write(chunk)' in str(e):
                for suffix in AUDIO_SUFFIX:
                    file_path = os.path.join(folder, file.replace(NCM_SUFFIX, suffix))
                    if os.path.exists(file_path):
                        os.remove(file_path)
                        print(f'\nDue to KeyboardInterrupt\n{file_path} is removed.')
            exit(0)
